### PR TITLE
Chapter 1 Solutions

### PR DESCRIPTION
**Pull request by siv12345**
### Solutions for Chapter {1}

cc @vrom911 @chshersh
 When I run the test suite
I get some errors in Task 5 where I have to find the number which is closer to zero from 2 negative numbers, but other than that it seems to be alright

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -508,7 +509,7 @@ branches because it is an expression and it must always return some value.
   satisfying the check will be returned and, therefore, evaluated.
 -}
 closestToZero :: Int -> Int -> Int
-closestToZero x y = error "closestToZero: not implemented!"
+closestToZero x y = min (abs x) (abs y)
```
Comment (chshersh) [inline src/Chapter1.hs:512]: This solution is almost correct :ok_hand: However, on negative numbers, it returns the absolute value of those negative numbers, note the value itself.

File: src/Chapter1.hs [hunk 2]
```diff
@@ -541,8 +542,11 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+  | x > y = mid y x z
+  | y > z = mid x z y
+  | otherwise = y
```
Comment (chshersh) [inline src/Chapter1.hs:549]: Short and clever solution!

File: src/Chapter1.hs [hunk 3]
```diff
@@ -619,9 +626,8 @@ Implement a function that returns the sum of the last two digits of a number.
 Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
-
-sumLast2 n = error "sumLast2: Not implemented!"
-
+sumLast2 :: Int -> Int
+sumLast2 n = (mod (div (abs n) 10) 10) + (mod (abs n) 10)
```
Comment (chshersh) [inline src/Chapter1.hs:630]: You were encouraged to use variables here to increase code readability :slightly_smiling_face: 

That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: the standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that is actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(first, second) = divMod (mod x 100) 10
```

You can see how we could pattern match on the pair 🙂

File: src/Chapter1.hs [hunk 4]
```diff
@@ -640,8 +646,10 @@ Implement a function that returns the first digit of a given number.
 You need to use recursion in this task. Feel free to return to it later, if you
 aren't ready for this boss yet!
 -}
-
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Int -> Int
+firstDigit n
+  | div n 10 == 0 = n
```
Comment (chshersh) [inline src/Chapter1.hs:651]: `div` on negative numbers returns negative number:

```haskell
ghci> div (-1) 10
-1
```

That's why this function hangs on CI and runs into an infinite loop. So, to make tests pass, you need to handle corner cases in this function :slightly_smiling_face:

**Review by chshersh — APPROVED** [review]
Congrats with finishing the first chapter!
And well done with the advanced task :+1: 

I left some comments and suggestions. Hope they help you to move forward :slightly_smiling_face:

**Comment by siv12345** [comment]
Thank you very much for the review of my PR @chshersh. I will make the changes you mentioned, locally on my machine. 

Can I close this PR?

**Comment by vrom911** [comment]
@siv12345 No, that is fine! Feel free to close this PR. The contributions would be counted toward Hacktoberfest 😉

**Comment by siv12345** [comment]
okay thanks, closing this pr
